devproxy/handlers/wurfl_handler/scientia_mobile_cloud.py

```diff
@@ -57,19 +84,47 @@ def get_device_from_smcloud(self, user_agent):
         """
         Queries ScientiaMobile's API and returns a dictionary of the device.
         """
-        # create basic auth string
+        # Create basic auth string
         b64 = base64.encodestring(self.smcloud_api_key).strip()
-        headers = {
-            #User-Agent is set by agent in getPage.
-            'X-Cloud-Client': self.SMCLOUD_CONFIG['client_version'],
-            'Authorization': 'Basic %s' % b64
-        }
-        try:
-            page = yield getPage(self.SMCLOUD_CONFIG['url'], headers=headers,
-                             agent=user_agent, timeout=5)
-        except ConnectError, exc:
-            raise ScientiaMobileCloudHandlerConnectError(exc)
-        device = json.loads(page)
+        if self.http_proxy_host:
+            headers = {
+                'X-Cloud-Client': [self.SMCLOUD_CONFIG['client_version']],
+                'Authorization': ['Basic %s' % b64]
+            }
+            if self.http_proxy_username and self.http_proxy_password:
+                auth = base64.encodestring(
+                    '%s:%s' % (
+                        self.http_proxy_username, self.http_proxy_password
+                    )
+                ).strip()
+                # Cater for many proxy servers
+                headers['Proxy-Authorization'] = ['Basic %s' % auth]
+                headers['Proxy-Authenticate'] = ['Basic %s' % auth]
+                headers['Proxy-Authentication'] = ['Basic %s' % auth]
+            endpoint = TCP4ClientEndpoint(
+                reactor, self.http_proxy_host, self.http_proxy_port or 80,
+                timeout=5
+            )
+            agent = ProxyAgent(endpoint)
+            qs = urlencode({'agent': user_agent})
+            response = yield agent.request('GET', self.SMCLOUD_CONFIG['url'] + '?' + qs,
+                headers=Headers(headers))
+            if response.code != 200:
+                raise ProxyConnectError()
+            d = Deferred()
+            response.deliverBody(SimpleReceiver(d))
+            body = yield d
+        else:
+            headers = {
+                'X-Cloud-Client': self.SMCLOUD_CONFIG['client_version'],
+                'Authorization': 'Basic %s' % b64
+            }
+            try:
+                body = yield getPage(self.SMCLOUD_CONFIG['url'], headers=headers,
+                    agent=user_agent, timeout=5)
+            except ConnectError, exc:
+                raise ScientiaMobileCloudHandlerConnectError(exc)
+        device = json.loads(body)
```

Comment:
Why have different mechanisms for fetching the page depending on whether we're using a proxy or not?

Rather factor the `ProxyAgent` construction out into a separate method and either call that or create a vanilla `Agent`. Then make the call to the WURFL with whichever agent you have and not worry about whether you're using a proxy or not at that point.